### PR TITLE
Add rest bff samples

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Add sample Ballerina and React code samples for backend to frontend(bff)

### CODE REVIEW

File: rest/backend_server/Ballerina.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[package]
+org = "bff"
+name = "rest"
+version = "0.1.0"
+distribution = "2201.8.2"
+
+[build-options]
```
Comment (LakshanWeerasinghe) [inline rest/backend_server/Ballerina.toml]: We can remove this if not needed.

File: rest/backend_server/server.bal [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    // Example: http://localhost:9090/sales/customers/C-124/orders?status=PENDING
+    resource function get customers/[string customerId]/orders(OrderStatus status = PENDING) returns Order[] {
+        return from Order entry in orders
+            where entry.customerId == customerId && entry.status == status
+            select entry;
+    };
+
+    // Example: http://localhost:9090/sales/orders
+    resource function post orders(Order orderRequest) returns Order|http:BadRequest {
+        if orders.hasKey(orderRequest.id) {
```
Comment (LakshanWeerasinghe) [inline rest/backend_server/server.bal]: Why the order id is not generated from the server side? Is it for make the use case more simple?

File: rest/backend_server/Ballerina.toml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+[package]
+org = "bff"
```
Comment (anupama-pathirage) [inline rest/backend_server/Ballerina.toml]: can we make it `bff_samples` to be aligned with integration samples? https://github.com/ballerina-guides/integration-samples/blob/main/github-pull-requests-to-gsheets/Ballerina.toml#L2

File: rest/backend_server/server.bal [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    resource function get orders() returns Order[] {
+        return orders.toArray();
+    };
+
+    // Example: http://localhost:9090/sales/orders/HM-238
+    resource function get orders/[string id]() returns Order|http:NotFound {
+        if orders.hasKey(id) {
+            return orders.get(id);
+        }
+        return <http:NotFound>{ body: string `Order not found. Order ID: ${id}` };
```
Comment (anupama-pathirage) [inline rest/backend_server/server.bal]: Not properly formatted

File: rest/backend_server/server.bal [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    resource function get customers/[string customerId]/orders(OrderStatus status = PENDING) returns Order[] {
+        return from Order entry in orders
+            where entry.customerId == customerId && entry.status == status
+            select entry;
+    };
+
+    // Example: http://localhost:9090/sales/orders
+    resource function post orders(Order orderRequest) returns Order|http:BadRequest {
+        if orders.hasKey(orderRequest.id) {
+            return <http:BadRequest>{ body: string `Order id already exists. Order ID: ${orderRequest.id}` };
```
Comment (anupama-pathirage) [inline rest/backend_server/server.bal]: Not properly formatted

File: rest/backend_server/server.bal (hunk 4 above)
Comment (anupama-pathirage) [inline rest/backend_server/server.bal]:
```suggestion
        return <http:NotFound>{body: string `Order not found. Order ID: ${id}`};
```

File: rest/backend_server/Ballerina.toml (hunk 3 above)
Comment (SasinduDilshara) [reply rest/backend_server/Ballerina.toml]: yes, I will update

File: rest/backend_server/Ballerina.toml (hunk 1 above)
Comment (SasinduDilshara) [reply rest/backend_server/Ballerina.toml]: Those are added while creating a new Ballerina project, 
Yes we can remove that.

File: rest/backend_server/server.bal (hunk 4 above)
Comment (SasinduDilshara) [reply rest/backend_server/server.bal]: I will update

File: rest/backend_server/server.bal (hunk 2 above)
Comment (SasinduDilshara) [reply rest/backend_server/server.bal]: It is done to focus on a `constraint validation` scenario and how to handle it in ballerina.`(http:BadRequest)`

File: rest/backend_server/Ballerina.toml (hunk 1 above)
Comment (SasinduDilshara) [reply rest/backend_server/Ballerina.toml]: updated. c15365d59e311c6748120203044dbfc6c4083a67

File: rest/backend_server/server.bal (hunk 4 above)
Comment (SasinduDilshara) [reply rest/backend_server/server.bal]: Updated. c15365d59e311c6748120203044dbfc6c4083a67

Comment (SasinduDilshara) [reply rest/backend_server/server.bal]: Updated. c15365d59e311c6748120203044dbfc6c4083a67

File: rest/backend_server/Ballerina.toml (hunk 3 above)
Comment (SasinduDilshara) [reply rest/backend_server/Ballerina.toml]: Updated. c15365d59e311c6748120203044dbfc6c4083a67

File: rest/backend_server/README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+## Ballerina with REST features
```
Comment (anupama-pathirage) [inline rest/backend_server/README.md]: Why using L2 heading here? It should L1

File: rest/backend_server/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+Most web apps rely on REST APIs, with JSON serving as the predominant data exchange format. Handling REST API elements like path parameters, query parameters, HTTP headers, status codes, and complex JSON structures is crucial for web app backends. Ballerina addresses these needs effectively by incorporating all these REST features as first-class citizens within the language itself, simplifying the backend development process and making it more intuitive and efficient.
+
+### Set up
```
Comment (anupama-pathirage) [inline rest/backend_server/README.md]: Why using L3 heading here? It should L2

**Comment by anupama-pathirage** [comment]
@SasinduDilshara Instead of just saying the sample name as `rest` can we have more descriptive sample name? Refer the sample names in https://github.com/ballerina-guides/integration-samples

Then reader can understand what we are going to do with rest here just by reading the sample name too. Have the same title in the README.md file of the sample.

**Comment by sameerajayasoma** [comment]
Maintain the `README.md` file inside the sample directory, not inside the `backend_server` directory?

File: rest/backend_server/README.md (hunk 6 above)
Comment (SasinduDilshara) [reply rest/backend_server/README.md]: updated

File: rest/backend_server/README.md (hunk 7 above)
Comment (SasinduDilshara) [reply rest/backend_server/README.md]: updated

**Comment by SasinduDilshara** [comment]
> Maintain the `README.md` file inside the sample directory, not inside the `backend_server` directory?

updated

**Comment by SasinduDilshara** [comment]
> @SasinduDilshara Instead of just saying the sample name as `rest` can we have more descriptive sample name? Refer the sample names in https://github.com/ballerina-guides/integration-samples
> 
> Then reader can understand what we are going to do with rest here just by reading the sample name too. Have the same title in the README.md file of the sample.

I added the name as `restful-order-service`.

File: restful-order-service/frontend_server/src/pages/index.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  const onStatusChange = (e) => {
+    setStatus(e.target.value);
+  }
+
+  return (
+    loading ? <div>Loading...</div> : error != null ? <div>{error}</div> :
+    <>
+      <Head>
+        <title>
+          orders | MegaPort Kit
```
Comment (sm1990) [inline restful-order-service/frontend_server/src/pages/index.js]:
```suggestion
          Orders | MegaPort Kit
```

File: restful-order-service/frontend_server/src/api/ApiHandler.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import axios from 'axios';
```
Comment (sm1990) [inline restful-order-service/frontend_server/src/api/ApiHandler.js:19]: Add the license agreement to all the js files inside the `src` folder.

File: restful-order-service/frontend_server/src/sections/order/orders-dropdown.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
```
Comment (sm1990) [inline restful-order-service/frontend_server/src/sections/order/orders-dropdown.js]: You can move the `orders` directory to the `components` directory.

Comment (SasinduDilshara) [reply restful-order-service/frontend_server/src/sections/order/orders-dropdown.js]: Will update the suggestions

Comment (SasinduDilshara) [reply restful-order-service/frontend_server/src/sections/order/orders-dropdown.js]: Updated. 05760c2213854b4a74ab82b6c9fab2797c59f922

**Review by sm1990 — DISMISSED** [review]

File: restful-order-service/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Restful Order Service with Ballerina
```
Comment (MaryamZi) [inline restful-order-service/README.md]: RESTful ?

File: restful-order-service/backend_server/Ballerina.toml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+[package]
+org = "bff_samples"
+name = "rest"
```
Comment (MaryamZi) [inline restful-order-service/backend_server/Ballerina.toml]: Shouldn't this also be updated?

**Review by sm1990 — DISMISSED** [review]

File: restful-order-service/backend_server/server.bal [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    resource function get orders() returns Order[] {
+        return orders.toArray();
+    };
+
+    // Example: http://localhost:9090/sales/orders/HM-238
+    resource function get orders/[string id]() returns Order|http:NotFound {
+        return orders[id] ?: <http:NotFound>{body: string `Order not found. Order ID: ${id}`};
+    };
+
+    // Example: http://localhost:9090/sales/customers/C-124/orders?status=PENDING
```
Comment (MaryamZi) [inline restful-order-service/backend_server/server.bal]: Since PENDING is the default, may be better to use a different example?

Comment (SasinduDilshara) [reply restful-order-service/backend_server/server.bal]: Updated

**Review by anupama-pathirage — APPROVED** [review]

**Pull request closed without merging**